cpp/src/arrow/util/cpu_info.cc

```diff
@@ -330,6 +346,29 @@ void CpuInfo::Init() {
   if (QueryPerformanceFrequency(&performance_frequency)) {
     max_mhz = static_cast<float>(performance_frequency.QuadPart);
   }
+#elif defined(__APPLE__)
+  // On macOS, get CPU information from system information base
+  struct SysCtlCpuFeature {
+    const char* name;
+    int64_t flag;
+  };
+  for (auto feature : std::vector<SysCtlCpuFeature>{
+           {"hw.optional.sse4_2", SSSE3 | SSE4_1 | SSE4_2 | POPCNT},
+           {"hw.optional.avx1_0", AVX},
+           {"hw.optional.avx2_0", AVX2},
+           {"hw.optional.bmi1", BMI1},
+           {"hw.optional.bmi2", BMI2},
+           {"hw.optional.avx512f", AVX512F},
+           {"hw.optional.avx512cd", AVX512CD},
+           {"hw.optional.avx512dq", AVX512DQ},
+           {"hw.optional.avx512bw", AVX512BW},
+           {"hw.optional.avx512vl", AVX512VL},
```

Comment:
Also add M1 arm64 flags?
`{"hw.optional.neon", ASIMD}`
I haven't tested this flag myself, just references https://github.com/golang/go/issues/42747#issuecomment-731208964

Reply:
Can do :-) By the way, is there a difference between "Neon" and "ASIMD"?

Reply:
Also, here are the available values on a M1:
https://gist.github.com/pitrou/fa1527038fea2d7a4cf367ecba7a2d2c

Reply:
Per my knowledge, `NEON` exists since 32 bit Arm, and `ASIMD` extends it to 64 bit system, with additional instructions and wider registers. `ASIMD` should be the correct term for Arm64 world, but obviously `NEON` is well known and widely used.

Reply:
Double checked with an expert from Arm. Actually **Neon == Advanced SIMD**. 

From Arm Architecture Reference Manual (ArmARM):
`The Advanced SIMD architecture, its associated implementations, and supporting software, are commonly referred
to as NEON™ technology". `

`Neon` was the codename that became public and more commonly used than the "proper name" (Advanced SIMD).